build_compilers/lib/src/dev_compiler_bootstrap_builder.dart

```diff
@@ -0,0 +1,272 @@
+        '"file":""}');
+  }
+
+  /// Ensures that all transitive js modules are available and built.
+  Future<Null> _ensureTransitiveModules(
+      Module module, AssetReader reader) async {
+    // Collect all the modules this module depends on, plus this module.
+    var transitiveDeps = await module.computeTransitiveDependencies(reader);
+    var jsModules = transitiveDeps
+        .map((id) => id.changeExtension(jsModuleExtension))
+        .toList()
+          ..add(module.jsId);
+    // Check that each module is readable, and warn otherwise.
+    await Future.wait(jsModules.map((jsId) async {
+      if (await reader.canRead(jsId)) return;
+      log.warning(
+          'Unable to read $jsId, check your console for compilation errors.');
+    }));
+  }
+
+  /// Returns whether or not [dartId] is an app entrypoint (basically, whether or
+  /// not it has a `main` function).
+  Future<bool> _isAppEntryPoint(AssetId dartId, AssetReader reader) async {
+    assert(dartId.extension == '.dart');
+    // Skip reporting errors here, dartdevc will report them later with nicer
+    // formatting.
+    var parsed = parseCompilationUnit(await reader.readAsString(dartId),
+        suppressErrors: true);
+    // Allow two or fewer arguments so that entrypoints intended for use with
+    // [spawnUri] get counted.
+    //
+    // TODO: This misses the case where a Dart file doesn't contain main(),
+    // but has a part that does, or it exports a `main` from another library.
+    return parsed.declarations.any((node) {
+      return node is FunctionDeclaration &&
+          node.name.name == "main" &&
+          node.functionExpression.parameters.parameters.length <= 2;
+    });
+  }
+}
+
+/// Code that actually imports the [moduleName] module, and calls the
+/// `[moduleScope].main()` function on it.
+///
+/// Also performs other necessary initialization.
+String _appBootstrap(String moduleName, String moduleScope) => '''
+require(["$moduleName", "dart_sdk"], function(app, dart_sdk) {
+  dart_sdk._isolate_helper.startRootIsolate(() => {}, []);
+$_initializeTools
+  app.$moduleScope.main();
+});
+})();
+''';
+
+/// The actual entrypoint JS file which injects all the necessary scripts to
+/// run the app.
+String _entryPointJs(String bootstrapModuleName) => '''
+(function() {
+  $_currentDirectoryScript
+  var el;
+  el = document.createElement("script");
+  el.defer = true;
+  el.async = false;
+  el.src = "/packages/\$sdk/dev_compiler/web/dart_stack_trace_mapper.js";
+  document.head.appendChild(el);
+  el = document.createElement("script");
+  el.defer = true;
+  el.async = false;
+  el.src = "/packages/\$sdk/dev_compiler/amd/require.js";
+  el.setAttribute("data-main", _currentDirectory + "$bootstrapModuleName");
+  document.head.appendChild(el);
+})();
+''';
+
+/// JavaScript snippet to determine the directory a script was run from.
+final _currentDirectoryScript = r'''
+var _currentDirectory = (function () {
```

Comment:
wow... this snippet has now been copied an impressive # of places. Maybe it should go in the DDC package.

Reply:
Ya, it probably should, as well as some of this other stuff. Not worth blocking on though imo.